iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/ipaddress/IpAddressValidator.java

```diff
@@ -0,0 +1,207 @@
+    ValidationResult validation = validateTrees();
+
+    if (validation.isValid()) {
+      long startLong = validation.startIp().toLong();
+      long endInt = validation.endIp().toLong();
+      for (Range safeRange : SAFE_RANGES) {
+        if (safeRange.contains(startLong) && safeRange.contains(endInt)) {
+          return;
+        }
+      }
+      reportIssue(ctx, message, secondaryLocationMessage);
+    }
+  }
+
+  private ValidationResult validateTrees() {
+    if (startIpAddress == null && endIpAddress == null) {
+      return new ValidationResult(false, null, null);
+    }
+    String start = null;
+    String end = null;
+    if (startIpAddress != null && startIpAddress.is(ArmTree.Kind.STRING_LITERAL)) {
+      start = TextUtils.getValue(startIpAddress).orElse(DEFAULT_START_IP);
+    }
+    if (endIpAddress != null && endIpAddress.is(ArmTree.Kind.STRING_LITERAL)) {
+      end = TextUtils.getValue(endIpAddress).orElse(DEFAULT_END_IP);
+    }
+    if (start == null && end == null) {
+      return new ValidationResult(false, null, null);
+    }
+    if (start == null) {
+      start = DEFAULT_START_IP;
+    }
+    if (end == null) {
+      end = DEFAULT_END_IP;
+    }
+
+    boolean isStartValid = Ip.isAddressValid(start);
+    boolean isEndValid = Ip.isAddressValid(end);
+    if (!isStartValid && !isEndValid) {
+      return new ValidationResult(false, null, null);
+    }
+    return new ValidationResult(
+      true,
+      isStartValid ? new Ip(start) : new Ip(DEFAULT_START_IP),
+      isEndValid ? new Ip(end) : new Ip(DEFAULT_END_IP));
+  }
+
+  private void reportIssue(CheckContext ctx, String message, String secondaryLocationMessage) {
+    Tree tree = (startIpAddress != null) ? startIpAddress : endIpAddress;
+    Tree secondary = (startIpAddress != null && endIpAddress != null) ? endIpAddress : null;
+
+    if (secondary != null) {
+      SecondaryLocation secondaryLocation = new SecondaryLocation(secondary, secondaryLocationMessage);
+      ctx.reportIssue(tree, message, List.of(secondaryLocation));
+    } else {
+      ctx.reportIssue(tree, message);
+    }
+  }
+
+  static class Ip {
+
+    private static final int BASE3 = (int) Math.pow(256, 3);
+    private static final int BASE2 = (int) Math.pow(256, 2);
+    private static final int BASE1 = 256;
+    private final String address;
+
+    Ip(String address) {
+      this.address = address;
+    }
+
+    static boolean isAddressValid(String text) {
+      // 0.0.0.0 - 7 chars, 255.255.255.255 - 15 chars
+      if (text.length() < 7 || text.length() > 15) {
+        return false;
+      }
+      String[] split = text.split("\\.");
+      if (split.length != 4) {
+        return false;
+      }
+      for (String s : split) {
+        try {
+          int x = Integer.parseInt(s);
+          if (x < 0 || x > 255) {
+            return false;
+          }
+        } catch (NumberFormatException e) {
+          return false;
+        }
+      }
+      return true;
```

Comment:
Alternatively, you could use the regex from OWASP:
`^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$`
It's a bit heavy and I'm not sure how fast/slow it is compared to your implementation.
It could also be used at the same time to capture each group.

Reply:
I'm sure that my impl will be faster - think about the whole state machine behind this regex.